### PR TITLE
[4.0] Upgrade inital pw length to 12 characters without complexitiy rules

### PR DESCRIPTION
**Pull request by zero-24**
Pull Request for Issue #29476

### Summary of Changes

As mentioned in a bunch of tickets (see https://github.com/joomla/joomla-cms/issues/29476), we have either nor password policy at all or a really bad one.

Therefore, we the JSST suggest the following for the Joomla Core:
* enforce a password policy in Joomla 4 during installation
* update the default settings of the password policy settings in com_users with the same policy

Policy would be:
* minimum 12 characters
* no other complexity rules

### Testing Instructions

- install this branch as new install
- choose an PW that is lower than 12 chars
- notice the message about the PW beeing to short.
- (please ignore the message about Mysql 8 as it is not realted to this changes and just come up when any kind of validation fails; this has to be fixed in a different PR)
- try to create an user from the frontend
- notice the minimum character count message
- try to create an user from the backend
- notice the minimum character count message

### Actual result BEFORE applying this Pull Request

by default in the installer we have no PW policy
by default we had the min 4 char PW policy.

### Expected result AFTER applying this Pull Request

by default we have a 12 character minimum lenth pw policy enfored in the installer and set per default 

### Documentation Changes Required

None.

### Other changes

- move language strings from com_users to the main as this is a field that should not depend on com_users strings
- upgrade the auto generated PW to 32 chars
- make the messages showing up as error given that the validation fails ;)

cc @joomla/security

### CODE REVIEW

File: libraries/src/Form/Rule/PasswordRule.php [hunk 1]
```diff
@@ -47,33 +47,38 @@ public function test(\SimpleXMLElement $element, $value, $group = null, Registry
...
 		$minimumUppercase = isset($element['minimum_uppercase']) ? (int) $element['minimum_uppercase'] : 0;
 		$minimumLowercase = isset($element['minimum_lowercase']) ? (int) $element['minimum_lowercase'] : 0;
 
-		// If we have parameters from com_users, use those instead.
-		// Some of these may be empty for legacy reasons.
-		$params = ComponentHelper::getParams('com_users');
-
-		if (!empty($params))
+		// Well at the installer we don't have any access to the
+		// database yet so lets us the hard coded default settings
```
Comment (brianteeman) [inline libraries/src/Form/Rule/PasswordRule.php]:
```suggestion
		// In the installer we don't have any access to the
		// database yet so use the hard coded default settings
```

Comment (zero-24) [reply libraries/src/Form/Rule/PasswordRule.php]: Thanks patched :)

**Comment by SharkyKZ** [comment]
Forcing 12 character minimum on all sites is unreasonable.

**Comment by zero-24** [comment]
> Forcing 12 character minimum on all sites is unreasonable.

Well it just the default configuration and can be changed at any time by the site owner.

That 12 character stuff is based on the recomendation of the US and EU authorities with us choosing not to enforce any complexity rules by default ad for that reasone choose 12 chars.

To me the initial super user has to be a secure PW where 12 chars is not optimal but a big step in the right direction in my view. (from just one char minumum right now)

**Comment by SharkyKZ** [comment]
No, it cannot be changed. 12 is now minimum allowed length. If this is not intentional, revert this change https://github.com/joomla/joomla-cms/pull/29859/files#diff-696886a9a37e99baabf37a9b31776d73L172-R172.

**Comment by zero-24** [comment]
Ha seems i missed that what kind of minimum there would you suggest 8 maybe?

**Comment by brianteeman** [comment]
NOTE : on upgraded sites you have to open and save the com_users options before the changes here have any effect

**Comment by zero-24** [comment]
Yes this is only for new installs right now. I don’t think we should mess with the settings of the users here right?

**Comment by Quy** [comment]
@zero-24 Don't make the same mistake as https://github.com/joomla/joomla-cms/pull/25177

**Comment by zero-24** [comment]
Yes right now i dont intend to update the config on upgraded sites.

**Comment by zero-24** [comment]
Patched thanks @SharkyKZ

File: administrator/components/com_users/config.xml [hunk 2]
```diff
@@ -169,10 +169,10 @@
 			name="minimum_length"
 			type="integer"
 			label="COM_USERS_CONFIG_FIELD_MINIMUM_PASSWORD_LENGTH"
-			first="4"
+			first="8"
```
Comment (PhilETaylor) [inline administrator/components/com_users/config.xml:172]: Suggest to also remove the 99 maximum minimum length too while you are at it and modifying this area. Its an artificial restriction. Its basically saying "You cannot demand your users to use 128char minimum passwords, but Joomla will let you demand 99 char minimum passwords from your users"... 99 is a purely made up figure.

File: administrator/components/com_users/config.xml [hunk 3]
```diff
@@ -169,10 +169,10 @@
...
 			last="99"
```
Comment (PhilETaylor) [inline administrator/components/com_users/config.xml]: Suggest to also remove the 99 maximum minimum length too while you are at it and modifying this area. Its an artificial restriction. Its basically saying "You cannot demand your users to use 128char minimum passwords, but Joomla will let you demand 99 char minimum passwords from your users"... 99 is a purely made up figure. Maybe use 256 and step it in 5s?

Comment (PhilETaylor) [reply administrator/components/com_users/config.xml]: And yes, I do appreciate the absurdity of a 256minimum length password ;-)

**Comment by PhilETaylor** [comment]
While I think no one is going to disagree with 

> enforce a password policy in Joomla 4 during installation [For the initial, first, Super Admin account]

I think most would disagree with 

> We have a 12 character minimum length and Site Owners are now no longer able to set a smaller figure for THEIR site AFTER installation using com_users configuration. Meaning EVERY USER EVER will need a 12 char minimum password. 

After installation those decisions should be made by the site admin, not the JSST. 

If a site admin wants 4 chars (like the government agency I worked for last month) then they should be allowed to be idiots and use 4 chars minimum. It should be configurable as it is now (although now they can go to 4 chars) - I dont know how low Joomla should "allow" but no one has ever complained at 4 chars.

Making it so every Joomla 4 user MUST have 12 char or above passwords for every user and not being able to override that as a Site Owner, will bring complaints. 

I can already see the blog posts explaining how to "fix" this "bug" to allow less chars on your site after installation... haha

Therefore I think increasing from 4 to 8 and allowing 8 chars (configurable down from 12 after installation) is a good compromise and can be spun as "better". 

-----

If I may suggest, and feel free to shoot it down, but a "development mode flag" such as that used for `JOOMLA_INSTALLATION_DISABLE_LOCALHOST_CHECK` and the skipping of the "delete installation folder to allow joomla access after install" so that DURING CODE DEVELOPMENT you can bypass the INSTALLER Password Policy ... 

Because "Aint no body got time for that", typing 12 chars, in development mode, instead of using admin/admin as we all do ;-)

**Comment by zero-24** [comment]
> If I may suggest, and feel free to shoot it down, but a "development mode flag" such as that used for `JOOMLA_INSTALLATION_DISABLE_LOCALHOST_CHECK` and the skipping of the "delete installation folder to allow joomla access after install" so that DURING CODE DEVELOPMENT you can bypass the INSTALLER Password Policy ... 
> 
> Because "Aint no body got time for that", typing 12 chars, in development mode, instead of using admin/admin as we all do ;-) Just a thought.

Well my main issue with that is that most of the localhost sites will be moved to a live site after developement right? So that would result into an super user without propper pw policy in public.

**Comment by PhilETaylor** [comment]
`JOOMLA_INSTALLATION_DISABLE_LOCALHOST_CHECK` can be used as an environment variable, like in the official docker images, so a similar thing could be used, and set as an environment variable locally. 

https://github.com/joomla/docker-joomla/blob/f57eb977eb1cdd6872c278be24458f281a9c92c3/php7.2/apache/Dockerfile#L5

File: administrator/components/com_users/config.xml (hunk 3 above)
Comment (PhilETaylor) [reply administrator/components/com_users/config.xml]: I see you removed the last=99 - this doesnt work and screws the dropdown :( 
Screenshot 2020-07-01 at 11 43 31

**Comment by zero-24** [comment]
> I dont think moving a site from dev to live would normally involve moving a ENV var too...

It is not about an env var. When you setup a local site with that env var set you would be allowed to bypass the PW rules and the user with the highest privileges would be setup. Now that site is ready and you move that to the live server the user is not changed and is still running with an PW not matching the pw rules.

I personally don't see a reason to allow admin/admin anyway. Given that all of us have many PW in mind anyway choosing an 12 char PW that you use for development should be possible, You can also use a PW manager for that.

> Because "Aint no body got time for that", typing 12 chars, in development mode, instead of using admin/admin as we all do ;-) Just a thought.

Well Please don't count me in on that `as we all do` I don't use admin as user name nor at any point admin as pw given that we all know that this is a well know combination and should not be used in my point of view.

**Comment by PhilETaylor** [comment]
It was just a thought - and you have provided good reasons why it should not happen. Im happy with that. 🎉🎉

**Comment by brianteeman** [comment]
> I think most would disagree with

>> We have a 12 character minimum length and Site Owners are now no longer able to set a smaller figure for THEIR site AFTER installation using com_users configuration. Meaning EVERY USER EVER will need a 12 char minimum password.

> After installation those decisions should be made by the site admin, not the JSST.

I 100% agree with enforcing a long password during installation for the super user.
I 100% agree with setting a long length as the default in com_users for new passwords

However a default setting is not the same as a minimum which is what is happening here. If I want my site to have a minimum of 9 or 10 characters I should be able to do that

**Comment by zero-24** [comment]
You are @brianteeman that was the thing @SharkyKZ noticed and was patched yesterday. Right now the minimum would be 8 up from 4

**Comment by PhilETaylor** [comment]
> However a default setting is not the same as a minimum which is what is happening here.

The new "default" is 12, but after installation you CAN downgrade that to 8 (previously it was 4) 

> If I want my site to have a minimum of 9 or 10 characters I should be able to do that

You can still after this

**Comment by PhilETaylor** [comment]
Well you could before https://github.com/joomla/joomla-cms/pull/29859/commits/5ff2fcb585a0f6a749480ffc0f75a1404340633e broke things - my bad for suggesting it ⚰️

**Comment by zero-24** [comment]
> Well you could before https://github.com/joomla/joomla-cms/pull/29859/commits/5ff2fcb585a0f6a749480ffc0f75a1404340633e broke things - my bad for suggesting it ⚰️
> 

What broke?

**Comment by PhilETaylor** [comment]
> What broke?

Commented here https://github.com/joomla/joomla-cms/pull/29859#discussion_r448277621

**Comment by brianteeman** [comment]
My bad I didnt update the pr

**Comment by PhilETaylor** [comment]
Might want to review https://github.com/joomla/joomla-cms/issues/29890 as well, because if the `data-min-length` was ACTUALLY being enforced then this PR would make it impossible for people to login after the User Min Length Options were saved if the new minimum was over the size of their password.

File: administrator/components/com_users/config.xml (hunk 3 above)
Comment (zero-24) [reply administrator/components/com_users/config.xml]: fixed i have set it to 2005 :D

**Comment by zero-24** [comment]
> Commented here #29859 (comment)

Fixed: https://github.com/joomla/joomla-cms/pull/29859/commits/7acdb6c9882e0d8e30476c14361cfb3a8b6778af

**Comment by PhilETaylor** [comment]
> Set the maximum length to 2005 (the year joomla was founded)

You might want to stop that by something else Joomla will be rendering 1997 rows of <options> increasing the page size considerably :) :) :

**Comment by zero-24** [comment]
> You might want to step that by something else Joomla will be rendering 1997 rows of increasing the page size considerably :) :) :

Well looks like there is a scroll bar in chrome? What browser do you use?
image

**Comment by PhilETaylor** [comment]
of course there is a scroll bar, in all browsers - but there is also 1997 rows of HTML to generate all those options :) 

`<option>1000</option>`

What is your total page size now? think slow networks, 3G, 3rd world internet speeds. 

Might be better just being a text box rather than a number selector?

**Comment by brianteeman** [comment]
Look at the source - it now has 1997 extra lines - just make it a regular text input - no one is going to scroll down to 2002

edit: typed at the same time as phil

**Comment by PhilETaylor** [comment]
Im actually not concerned people will need to scroll - Im more concerned at the overall page size in Mb that needs to be downloaded. Im mobile so I cannot check the actual size, but from experience (when I screwed up), having an extra 2000 lines of HTML for one dropdown (or in my case 30 dropdown of 1000 rows) the page size was huge. 

I may have 1Gbps symmetrical fibre to the desk - but others in the world only have a "smaller" connection to the internet :)

**Comment by brianteeman** [comment]
I was writing in shorthand - file size is absolutely an issue

**Comment by zero-24** [comment]
Please check and test the latest commit that should adress the issue and even removing an suggested maximum at this place.

File: administrator/components/com_users/config.xml [hunk 4]
```diff
@@ -147,72 +147,74 @@
...
 		<field
 			name="reset_count"
-			type="integer"
+			type="number"
 			label="COM_USERS_CONFIG_FIELD_FRONTEND_RESET_COUNT_LABEL"
-			first="0"
-			last="20"
+			min="0"
+			max="20"
 			step="1"
```
Comment (Quy) [inline administrator/components/com_users/config.xml:154]: Remove. Repeat below.

Comment (zero-24) [reply administrator/components/com_users/config.xml:154]: Why? there is a step parameter in the number field? https://github.com/joomla/joomla-cms/blob/98c868f10a6ba76b0e29bfb4967bb9b935ff37ff/libraries/src/Form/Field/NumberField.php#L58

Comment (Quy) [reply administrator/components/com_users/config.xml:154]: Sorry nevermind.

**Comment by Quy** [comment]
Remove.

https://github.com/zero-24/joomla-cms/blob/86f2247aa157062647d3763a62c337f423ba89ef/language/en-GB/com_users.ini#L50-L60

**Comment by zero-24** [comment]
Why? You can still configure such rule?

**Comment by Quy** [comment]
Those strings are obsolete because of the new strings.

**Comment by zero-24** [comment]
Patched thanks @Quy

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 6c4e182d7b40c01100c2934045a2a05d1cc0246f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29859">issues.joomla.org/tracker/joomla-cms/29859</a>.</sub>

**Comment by jmeintrup** [comment]
I have tested this item :white_check_mark: successfully on 6c4e182d7b40c01100c2934045a2a05d1cc0246f<br /><br />Tested successfully


One note: I think it's too late to inform the user during setup about their password choice AFTER the database form is filled out.
It should be right after the admin password is set, similar to a message about a missing mandatory field.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29859">issues.joomla.org/tracker/joomla-cms/29859</a>.</sub>

**Comment by david0296** [comment]
I have tested this item :white_check_mark: successfully on 6c4e182d7b40c01100c2934045a2a05d1cc0246f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29859">issues.joomla.org/tracker/joomla-cms/29859</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29859">issues.joomla.org/tracker/joomla-cms/29859</a>.</sub>

File: administrator/components/com_users/config.xml (hunk 3 above)
Comment (wilsonge) [reply administrator/components/com_users/config.xml]: This is actually too high already. Bcrypt has an actual max length https://www.mscharhag.com/software-development/bcrypt-maximum-password-length

Comment (PhilETaylor) [reply administrator/components/com_users/config.xml]: Another good reason NOT to use Bcrypt then. 

Symfony has a hard coded 4096 limit on password length ([CVE-2013-5750: Security issue](https://symfony.com/blog/cve-2013-5750-security-issue-in-fosuserbundle-login-form))

**Comment by Quy** [comment]
Should these be `12`?

https://github.com/joomla/joomla-cms/blob/79ead2cd204dab8b148868b9c83005f647055754/libraries/src/Form/Field/PasswordField.php#L157

https://github.com/joomla/joomla-cms/blob/79ead2cd204dab8b148868b9c83005f647055754/libraries/src/Form/Field/PasswordField.php#L165

On installation page, password and db password are `4`. Changing to `12` will force both to be `12`.
```
<input type="password" name="jform[admin_password]" id="jform_admin_password" value="" autocomplete="new-password" class="form-control form-control required" maxlength="99" required="" data-min-length="4">
```

**Comment by zero-24** [comment]
Yes please do a PR seems i have missed them when checking the places where we used 4 before.

**Comment by Quy** [comment]
So OK to force db password to be a minimum of 12?

**Comment by zero-24** [comment]
Well the db Passwort is out of our control we can not force any rule there.

**Comment by infograf768** [comment]
> So OK to force db password to be a minimum of 12?

I sincerely hope not as we would not be able to use `root` anymore.

**Comment by brianteeman** [comment]
Please remember that on many hosting environments the mysql password is provided by the host and is not under the users control

**Comment by Quy** [comment]
There is `data-min-length` attribute in db password but no validation is applied.

`<input type="password" name="jform[db_pass]" id="jform_db_pass" value="" autocomplete="off" class="form-control form-control" maxlength="99" data-min-length="12">`

**Comment by zero-24** [comment]
> There is data-min-length attribute in db password but no validation is applied.

yes and we can not enforce any validation given that this is out of our and in most cases also the site owners control. we might want to remove that maxlength and min length attributes from the output.

**Pull request merged**